src/sparsematrix.jl

```diff
@@ -748,6 +748,40 @@ SparseMatrixCSC{Tv}(M::Transpose{<:Any,<:AbstractSparseMatrixCSC}) where {Tv} =
 SparseMatrixCSC{Tv,Ti}(M::Adjoint{<:Any,<:AbstractSparseMatrixCSC}) where {Tv,Ti} = SparseMatrixCSC{Tv,Ti}(copy(M))
 SparseMatrixCSC{Tv,Ti}(M::Transpose{<:Any,<:AbstractSparseMatrixCSC}) where {Tv,Ti} = SparseMatrixCSC{Tv,Ti}(copy(M))
 
+function SparseMatrixCSC_fromcol(Tv, Ti, M; sizehint=false)
+    
+    colptr = zeros(Ti, size(M, 2) + 1)
+    nzval = Vector{Tv}(undef, 0)
+    rowval = Vector{Ti}(undef, 0)
+    if sizehint
+        nz = 0
+        for col in eachcol(M)
+            for i in col
+                if iszero(i) == false
+                    nz += 1
+                end
+            end
+        end
+        sizehint!(nzval, nz)
+        sizehint!(rowval, nz)
+    end
+    colptr[1] = 1
+    ind = 1
+    for (j, col) in enumerate(eachcol(M))
+        for (i, v) in enumerate(col)
+            if iszero(v) == false
+                push!(nzval, v)
+                push!(rowval, i)
+                ind += 1
+            end
+        end
+        colptr[j + 1] = ind
+    end
```

Comment:
Doesn't that still index into each column of `Q` element per element? I guess what you want is something like
```suggestion
    for (j, colind) in enumerate(axes(M, 2))
        col = M[:, colind] # this could be further optimized by allocating col once and lmul! with Q
        for i in axes(v, 1)
            if !iszero(v) # should be _isnotzero(v)
                push!(nzval, v)
                push!(rowval, i)
                ind += 1
            end
        end
        colptr[j + 1] = ind
    end
```

Reply:
it does what i want already but i guess i can do  `size(Q, 2) - 1` fewer allocations. by accessing it first with `eachcol(M)` then on each element, i'm not doing a getindex (which loads the whole column) for every element.

Reply:
The elements in the iterator of eachcol are views into the columns, but remember that `Q` is not stored as a matrix, the columns are computed by lmul!-ing the canonical vectors. So I don't see how it should know to materialize that view via `getindex(Q, :, j)`.

Reply:
@dkarrasch i've implemented it like you said, it's faster but i don't understand why it's faster. As far i as understand, going over columns should only call lmul! once per column, right? or 

```
julia> @btime SparseArrays._from_lmul(Float64, Int, $q.Q);
  74.427 μs (3991 allocations: 197.88 KiB)

julia> @btime SparseArrays._from_col(Float64, Int, $q.Q);
  5.616 ms (390190 allocations: 25.96 MiB)
```

Reply:
`eachcol` does not call `getindex(Q, :, j)` for `j in ...`, which would make use of that `lmul!` "trick". It just creates a view into the columns, which are thin wrappers, no "colwise" magic behind the scenes. So, when you start to iterate over a view into one column, it ends up with scalar indexing into the parent array, which is exactly what you want to avoid.

Reply:
BTW, I would restrict that "`SparseMatrixCSC`" method to `AbstractQ`s directly. For matrices that are materialized in memory, you don't even want to consider this. Same for structured matrices, which most likely already have their own `SparseMatrixCSC` constructor.

Reply:
i've already done that no? 
```julia
SparseMatrixCSC{Tv, Ti}(Q::LinearAlgebra.AbstractQ) where {Tv, Ti} = _from_lmul(Tv, Ti, Q; sizehint=false)
```
i think putting a type on _from_lmul/_from_eachcol from makes it too narrow. i think the name is enough?

BTW. should i make a patch for `eachcol(AbstractQ)` or is it over specialization?

Reply:
Sorry, I might have missed the (then) latest changes. For experimenting its fine to have it like you have now, but eventually I would move the long function body right under the `SparseMatrixCSC` header, that's what I meant.

> BTW. should i make a patch for eachcol(AbstractQ) or is it over specialization?

The underlying problem is that `AbstratQ`s are not really matrices, and the expectation is that `eachcol` is just a thin iterator wrapper, not something that comes with a potentially significant computation price. I'm experimenting locally with removing the subtyping `AbstractQ <: AbstractMatrix`, because that has caused other issues in the past.

Reply:
> SparseMatrixCSC header, that's what I meant.
done i think.

i feel like removing the AbstractMatrix is going to break a bunch of stuff (that have poor performance) in random packages. I think having an `eachcol` is the right balance between writing complicated stuff and performance. i.e.
```julia
Base.eachcol(Q::AbstractQ) = (Q[:, i] for i in axes(Q, 2))
```

Reply:
But we're not using `eachcol(::AbstractQ)`, are we? The proposal is a good, but it seems to me that it doesn't follow the `eachcol` logic, it behaves differently. The docstring says that it returns views into the columns, so in particular you can write into them and thereby modify the underlying array, whic is not the case here. I'd suggest to avoid adding anything that is not really necessary for the current issue.